File: import-tsv.py
#!/usr/bin/env python3
import os
import sys
import glob
import psycopg2
import argparse
from datetime import datetime


def parse_metadata(lines):
    """Extract metadata from header lines"""
    metadata = {
        'software_version': '',
        'database_version': '',
        'database_type': '',
        'doi': '',
        'url': ''
    }
    
    for line in lines:
        if line.startswith('# Software:'):
            metadata['software_version'] = line.replace('# Software:', '').strip()
        elif line.startswith('# Database:'):
            parts = line.replace('# Database:', '').strip().split(',')
            if len(parts) >= 1:
                metadata['database_version'] = parts[0].strip()
            if len(parts) >= 2:
                metadata['database_type'] = parts[1].strip()
        elif line.startswith('# DOI:'):
            metadata['doi'] = line.replace('# DOI:', '').strip()
        elif line.startswith('# URL:'):
            metadata['url'] = line.replace('# URL:', '').strip()
    
    return metadata


def parse_tsv_file(file_path):
    """Parse a Bakta TSV file and return metadata and annotations"""
    with open(file_path, 'r') as f:
        lines = f.readlines()
    
    # Separate header and data lines
    header_lines = [line for line in lines if line.startswith('#')]
    data_lines = [line for line in lines if not line.startswith('#')]
    
    # Extract sample_id from filename (without extension)
    sample_id = os.path.splitext(os.path.basename(file_path))[0]
    
    # Parse metadata from header
    metadata = parse_metadata(header_lines)
    metadata['sample_id'] = sample_id
    metadata['file_path'] = file_path
    
    # Parse annotations
    annotations = []
    for line in data_lines:
        if not line.strip():
            continue
            
        parts = line.strip().split('\t')
        if len(parts) < 8:  # Ensure we have at least the main fields
            continue
            
        annotation = {
            'sequence_id': parts[0],
            'feature_type': parts[1],
            'start_position': int(parts[2]),
            'stop_position': int(parts[3]),
            'strand': parts[4],
            'locus_tag': parts[5] if parts[5] else None,
            'gene': parts[6] if parts[6] else None,
            'product': parts[7] if parts[7] else None,
            'dbxrefs': parts[8].split(', ') if len(parts) > 8 and parts[8] else []
        }
        annotations.append(annotation)
    
    return metadata, annotations


def import_to_database(conn, metadata, annotations):
    """Import metadata and annotations to the PostgreSQL database"""
    cursor = conn.cursor()
    
    try:
        # Start transaction
        cursor.execute("BEGIN")
        
        # Insert genome data and get its ID
        cursor.execute("""
            INSERT INTO genomes (sample_id, software_version, database_version, database_type, doi, url, file_path)
            VALUES (%s, %s, %s, %s, %s, %s, %s)
            RETURNING id
        """, (
            metadata['sample_id'],
            metadata['software_version'],
            metadata['database_version'],
            metadata['database_type'],
            metadata['doi'],
            metadata['url'],
            metadata['file_path']
        ))
        
        genome_id = cursor.fetchone()[0]
        
        # Insert annotations in batches for efficiency
        batch_size = 1000
        for i in range(0, len(annotations), batch_size):
            batch = annotations[i:i+batch_size]
            
            values_str = []
            values_list = []
            
            for annot in batch:
                values_str.append("(%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)")
                values_list.extend([
                    genome_id,
                    annot['sequence_id'],
                    annot['feature_type'],
                    annot['start_position'],
                    annot['stop_position'],
                    annot['strand'],
                    annot['locus_tag'],
                    annot['gene'],
                    annot['product'],
                    annot['dbxrefs'] if annot['dbxrefs'] else None
                ])
            
            query = f"""
                INSERT INTO annotations 
                (genome_id, sequence_id, feature_type, start_position, stop_position, 
                strand, locus_tag, gene, product, dbxrefs)
                VALUES {', '.join(values_str)}
            """
            
            cursor.execute(query, values_list)
        
        # Commit the transaction
        cursor.execute("COMMIT")
        
    except Exception as e:
        cursor.execute("ROLLBACK")
        print(f"Error during database import: {e}")
        raise


def main():
    parser = argparse.ArgumentParser(description='Import Bakta annotation TSV files into PostgreSQL database')
    parser.add_argument('--folder', help='Folder containing Bakta TSV files')
    parser.add_argument('--db-name', default='bakta_annotations', help='PostgreSQL database name')
    parser.add_argument('--db-user', default='postgres', help='PostgreSQL username')
    parser.add_argument('--db-password', help='PostgreSQL password')
    parser.add_argument('--db-host', default='localhost', help='PostgreSQL host')
    parser.add_argument('--db-port', default='5432', help='PostgreSQL port')
    parser.add_argument('--pattern', default='*.tsv', help='File pattern to match TSV files')
    
    args = parser.parse_args()
    
    # Connect to PostgreSQL database
    try:
        conn = psycopg2.connect(
            dbname=args.db_name,
            user=args.db_user,
            password=args.db_password,
            host=args.db_host,
            port=args.db_port
        )
    except Exception as e:
        print(f"Error connecting to database: {e}")
        sys.exit(1)
    
    # Find all TSV files in the specified folder
    folder_path = os.path.abspath(args.folder)
    file_pattern = os.path.join(folder_path, args.pattern)
    tsv_files = glob.glob(file_pattern)
    
    if not tsv_files:
        print(f"No TSV files found in {folder_path} matching pattern {args.pattern}")
        sys.exit(1)
    
    print(f"Found {len(tsv_files)} TSV files to import")
    
    # Process each file
    success_count = 0
    error_count = 0
    
    for i, file_path in enumerate(tsv_files, 1):
        try:
            print(f"[{i}/{len(tsv_files)}] Processing {os.path.basename(file_path)}")
            metadata, annotations = parse_tsv_file(file_path)
            import_to_database(conn, metadata, annotations)
            print(f"  Imported {len(annotations)} annotations")
            success_count += 1
        except Exception as e:
            print(f"  Error processing {file_path}: {e}")
            error_count += 1
    
    conn.close()
    print(f"Import completed. Successfully processed {success_count} files. Failed: {error_count} files.")


if __name__ == "__main__":
    main()

File: import-tsv.mjs
#!/usr/bin/env node

import fs from 'fs/promises';
import path from 'path';
import pg from 'pg';
import { fileURLToPath } from 'url';

// Command line argument handling
const args = process.argv.slice(2);
if (args.length === 0 || args[0] === '-h' || args[0] === '--help') {
  console.log(`
Usage: node import_bakta.mjs <folder_path> [options]

Options:
  --db-name       PostgreSQL database name (default: bakta_tsv)
  --db-user       PostgreSQL username (default: postgres)
  --db-password   PostgreSQL password
  --db-host       PostgreSQL host (default: localhost)
  --db-port       PostgreSQL port (default: 5432)
  --pattern       File pattern to match TSV files (default: .tsv)

Example:
  node import_bakta.mjs ./data_folder --db-password=mypassword
`);
  process.exit(0);
}

// Parse command line arguments
const folderPath = args[0];
let dbName = 'bakta_tsv';
let dbUser = 'postgres';
let dbPassword = '';
let dbHost = 'localhost';
let dbPort = 5432;
let filePattern = '.tsv';

for (let i = 1; i < args.length; i++) {
  const arg = args[i];
  if (arg.startsWith('--db-name=')) {
    dbName = arg.replace('--db-name=', '');
  } else if (arg.startsWith('--db-user=')) {
    dbUser = arg.replace('--db-user=', '');
  } else if (arg.startsWith('--db-password=')) {
    dbPassword = arg.replace('--db-password=', '');
  } else if (arg.startsWith('--db-host=')) {
    dbHost = arg.replace('--db-host=', '');
  } else if (arg.startsWith('--db-port=')) {
    dbPort = arg.replace('--db-port=', '');
  } else if (arg.startsWith('--pattern=')) {
    filePattern = arg.replace('--pattern=', '');
  }
}

// Connect to PostgreSQL
const pool = new pg.Pool({
  user: dbUser,
  host: dbHost,
  database: dbName,
  password: dbPassword,
  port: dbPort,
});

/**
 * Parse metadata from header lines
 * @param {string[]} lines - Header lines from the TSV file
 * @returns {Object} - Metadata object
 */
function parseMetadata(lines) {
  const metadata = {
    software_version: '',
    database_version: '',
    database_type: '',
    doi: '',
    url: ''
  };

  for (const line of lines) {
    if (line.startsWith('# Software:')) {
      metadata.software_version = line.replace('# Software:', '').trim();
    } else if (line.startsWith('# Database:')) {
      const parts = line.replace('# Database:', '').trim().split(',');
      if (parts.length >= 1) {
        metadata.database_version = parts[0].trim();
      }
      if (parts.length >= 2) {
        metadata.database_type = parts[1].trim();
      }
    } else if (line.startsWith('# DOI:')) {
      metadata.doi = line.replace('# DOI:', '').trim();
    } else if (line.startsWith('# URL:')) {
      metadata.url = line.replace('# URL:', '').trim();
    }
  }

  return metadata;
}

/**
 * Process a single TSV file and import it into the database
 * @param {string} filePath - Path to the TSV file
 */
async function processTsvFile(filePath) {
  console.log(`Processing file: ${filePath}`);
  
  const client = await pool.connect();
  
  try {
    await client.query('BEGIN');
    
    // Read and parse file
    const fileContent = await fs.readFile(filePath, 'utf-8');
    const lines = fileContent.split('\n');
    
    // Separate header and data lines
    const headerLines = lines.filter(line => line.startsWith('#'));
    const dataLines = lines.filter(line => !line.startsWith('#') && line.trim() !== '');
    
    // Extract metadata from header
    const metadata = parseMetadata(headerLines);
    
    // Use filename as sample_id
    const fileName = path.basename(filePath);
    const sampleId = fileName.replace(/\.[^/.]+$/, ''); // Remove file extension
    
    // Insert genome data
    const genomeResult = await client.query(`
      INSERT INTO genomes (sample_id, software_version, database_version, database_type, doi, url, file_path)
      VALUES ($1, $2, $3, $4, $5, $6, $7)
      RETURNING id
    `, [
      sampleId,
      metadata.software_version,
      metadata.database_version,
      metadata.database_type,
      metadata.doi,
      metadata.url,
      filePath
    ]);
    
    const genomeId = genomeResult.rows[0].id;
    
    // Process annotations in batches
    const batchSize = 1000;
    const totalBatches = Math.ceil(dataLines.length / batchSize);
    
    for (let batchIndex = 0; batchIndex < totalBatches; batchIndex++) {
      const start = batchIndex * batchSize;
      const end = Math.min(start + batchSize, dataLines.length);
      const batch = dataLines.slice(start, end);
      
      // Prepare values for batch insert
      const values = [];
      const valueStrings = [];
      let valueIndex = 1;
      
      for (const line of batch) {
        const parts = line.split('\t');
        if (parts.length < 8) continue;
        
        // Parse dbxrefs if available
        const dbxrefs = parts.length > 8 && parts[8] 
          ? parts[8].split(', ').filter(x => x.trim() !== '') 
          : null;
        
        valueStrings.push(`($${valueIndex++}, $${valueIndex++}, $${valueIndex++}, $${valueIndex++}, $${valueIndex++}, $${valueIndex++}, $${valueIndex++}, $${valueIndex++}, $${valueIndex++}, $${valueIndex++})`);
        
        values.push(
          genomeId,
          parts[0], // sequence_id
          parts[1], // feature_type
          parseInt(parts[2]), // start_position
          parseInt(parts[3]), // stop_position
          parts[4], // strand
          parts[5] || null, // locus_tag
          parts[6] || null, // gene
          parts[7] || null, // product
          dbxrefs // dbxrefs
        );
      }
      
      if (valueStrings.length > 0) {
        const query = `
          INSERT INTO annotations 
          (genome_id, sequence_id, feature_type, start_position, stop_position, strand, locus_tag, gene, product, dbxrefs)
          VALUES ${valueStrings.join(', ')}
        `;
        
        await client.query(query, values);
      }
    }
    
    await client.query('COMMIT');
    console.log(`Imported ${dataLines.length} annotations from ${fileName}`);
    
  } catch (error) {
    await client.query('ROLLBACK');
    console.error(`Error processing file ${filePath}:`, error);
    throw error;
  } finally {
    client.release();
  }
}

/**
 * Main function to process all TSV files in a directory
 */
async function main() {
  try {
    // Get all files in the directory
    const files = await fs.readdir(folderPath);
    
    // Filter for TSV files
    const tsvFiles = files.filter(file => file.toLowerCase().includes(filePattern));
    
    if (tsvFiles.length === 0) {
      console.log(`No TSV files found in ${folderPath} matching pattern ${filePattern}`);
      process.exit(1);
    }
    
    console.log(`Found ${tsvFiles.length} TSV files to import`);
    
    // Process each file sequentially to avoid overwhelming the database
    for (let i = 0; i < tsvFiles.length; i++) {
      const filePath = path.join(folderPath, tsvFiles[i]);
      await processTsvFile(filePath);
      console.log(`Processed ${i + 1}/${tsvFiles.length} files`);
    }
    
    console.log('All files processed successfully');
    await pool.end();
    
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

// Run the main function
main();